[Quant][PT2E] Add cat and avg_pool2d recipe into x86InductorQuantizer

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107951
* #106958
* #106838
* __->__ #106836

**Summary**
Add `cat` and `avg_pool2d` quantization recipe as input output share observer into `x86InductorQuantizer`.

**Test Plan**
```
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_same_inputs
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_single_input
clear && python -m pytest test_x86inductor_quantizer.py -k test_avg_pool2d_recipe
```

leslie-fang-intel committed:
add cat and avg_pool2d recipe into x86InductorQuantizer

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Add cat and avg_pool2d recipe into x86InductorQuantizer"


**Summary**
Add `cat` and `avg_pool2d` quantization recipe as input output share observer into `x86InductorQuantizer`.

**Test Plan**
```
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_same_input
clear && python -m pytest test_x86inductor_quantizer.py -k test_avg_pool2d_recipe
```



[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Add cat and avg_pool2d recipe into x86InductorQuantizer"


**Summary**
Add `cat` and `avg_pool2d` quantization recipe as input output share observer into `x86InductorQuantizer`.

**Test Plan**
```
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_same_inputs
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_single_input
clear && python -m pytest test_x86inductor_quantizer.py -k test_avg_pool2d_recipe
```



[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Add cat and avg_pool2d recipe into x86InductorQuantizer"


**Summary**
Add `cat` and `avg_pool2d` quantization recipe as input output share observer into `x86InductorQuantizer`.

**Test Plan**
```
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_same_inputs
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_single_input
clear && python -m pytest test_x86inductor_quantizer.py -k test_avg_pool2d_recipe
```



[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Add cat and avg_pool2d recipe into x86InductorQuantizer"


**Summary**
Add `cat` and `avg_pool2d` quantization recipe as input output share observer into `x86InductorQuantizer`.

**Test Plan**
```
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_same_inputs
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_single_input
clear && python -m pytest test_x86inductor_quantizer.py -k test_avg_pool2d_recipe
```



[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Add cat and avg_pool2d recipe into x86InductorQuantizer"


**Summary**
Add `cat` and `avg_pool2d` quantization recipe as input output share observer into `x86InductorQuantizer`.

**Test Plan**
```
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_same_inputs
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_single_input
clear && python -m pytest test_x86inductor_quantizer.py -k test_avg_pool2d_recipe
```



[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Add cat and avg_pool2d recipe into x86InductorQuantizer"


**Summary**
Add `cat` and `avg_pool2d` quantization recipe as input output share observer into `x86InductorQuantizer`.

**Test Plan**
```
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_same_inputs
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_single_input
clear && python -m pytest test_x86inductor_quantizer.py -k test_avg_pool2d_recipe
```



[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Add cat and avg_pool2d recipe into x86InductorQuantizer"


**Summary**
Add `cat` and `avg_pool2d` quantization recipe as input output share observer into `x86InductorQuantizer`.

**Test Plan**
```
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_same_inputs
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_single_input
clear && python -m pytest test_x86inductor_quantizer.py -k test_avg_pool2d_recipe
```



[ghstack-poisoned]

leslie-fang-intel commented:
@jerryzh168 @jgong5 Could you help to take a look of this PR?

jgong5 commented on torch/ao/quantization/quantizer/x86_inductor_quantizer.py:
```diff
@@ -535,6 +540,36 @@ def _annotate_maxpool2d(
+            if input_node not in input_qspec_map:
+                # There has the case of cat same nodes: torch.cat([input0, input0], 1)
+                assert isinstance(input_node, Node)
+                assert isinstance(
+                    share_qparams_with_input_act0_qspec, QuantizationSpecBase
+                )
```
`share_qparams_with_input_act0_qspec` is constructed as `SharedQuantizationSpec`. This check seems redundant?

leslie-fang-intel replied:
I remember there are some format check error if we remove it previously. I will remove it and check the preCI again.

jgong5 commented on torch/ao/quantization/quantizer/x86_inductor_quantizer.py:
```diff
@@ -595,10 +650,48 @@ def _annotate_output_for_int8_in_int8_out_pattern(
+            elif node.target is torch.ops.aten.cat.default:
+                cat_node = node
+                cat_quantization_annotation = (
+                    cat_node.meta["quantization_annotation"]
```
Is it better to build a constant variable for the key here?

leslie-fang-intel replied:
Not sure if my understanding is right or not. Did you mean define `cat_quantization_annotation 
 = cat_node.meta["quantization_annotation"]` and use `cat_quantization_annotation` later instead of `cat_node.meta["quantization_annotation"]`?

jgong5 replied:
Oh, I mean using some constant variable like `QUANT_ANNOTATION_KEY` instead of explicitly using `"quantization_annotation"` everywhere.

leslie-fang-intel replied:
Sure, make sense and changed. Please help to take a look again.

jgong5 commented on torch/ao/quantization/quantizer/x86_inductor_quantizer.py:
```diff
@@ -595,10 +650,48 @@ def _annotate_output_for_int8_in_int8_out_pattern(
+                    avg_pool2d_quantization_annotation
+                    and avg_pool2d_quantization_annotation._is_output_of_quantized_pattern
+                ):
+                    # Annotate the output_qspec of cat_node
```
This comment is not aligned with avg_pool2d you are handling. Also, is it possible to factor out the common code for annotating ops with `SharedQuantizationSpec`?

leslie-fang-intel replied:
Thanks for comment. Yeap, refactor the common code for `cat` and `avg_pool2d`. As for `max_pool2d_with_indices`, keep it as it since it's slightly different with 2 nodes in the pattern actually.

leslie-fang-intel committed:
Update on "[Quant][PT2E] Add cat and avg_pool2d recipe into x86InductorQuantizer"


**Summary**
Add `cat` and `avg_pool2d` quantization recipe as input output share observer into `x86InductorQuantizer`.

**Test Plan**
```
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_same_inputs
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_single_input
clear && python -m pytest test_x86inductor_quantizer.py -k test_avg_pool2d_recipe
```



[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Add cat and avg_pool2d recipe into x86InductorQuantizer"


**Summary**
Add `cat` and `avg_pool2d` quantization recipe as input output share observer into `x86InductorQuantizer`.

**Test Plan**
```
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_same_inputs
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_single_input
clear && python -m pytest test_x86inductor_quantizer.py -k test_avg_pool2d_recipe
```



[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Add cat and avg_pool2d recipe into x86InductorQuantizer"


**Summary**
Add `cat` and `avg_pool2d` quantization recipe as input output share observer into `x86InductorQuantizer`.

**Test Plan**
```
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_same_inputs
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_single_input
clear && python -m pytest test_x86inductor_quantizer.py -k test_avg_pool2d_recipe
```



[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Add cat and avg_pool2d recipe into x86InductorQuantizer"


**Summary**
Add `cat` and `avg_pool2d` quantization recipe as input output share observer into `x86InductorQuantizer`.

**Test Plan**
```
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_same_inputs
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_single_input
clear && python -m pytest test_x86inductor_quantizer.py -k test_avg_pool2d_recipe
```



[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Add cat and avg_pool2d recipe into x86InductorQuantizer"


**Summary**
Add `cat` and `avg_pool2d` quantization recipe as input output share observer into `x86InductorQuantizer`.

**Test Plan**
```
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_same_inputs
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_single_input
clear && python -m pytest test_x86inductor_quantizer.py -k test_avg_pool2d_recipe
```



[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Add cat and avg_pool2d recipe into x86InductorQuantizer"


**Summary**
Add `cat` and `avg_pool2d` quantization recipe as input output share observer into `x86InductorQuantizer`.

**Test Plan**
```
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_same_inputs
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_single_input
clear && python -m pytest test_x86inductor_quantizer.py -k test_avg_pool2d_recipe
```



[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Add cat and avg_pool2d recipe into x86InductorQuantizer"


**Summary**
Add `cat` and `avg_pool2d` quantization recipe as input output share observer into `x86InductorQuantizer`.

**Test Plan**
```
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_same_inputs
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_single_input
clear && python -m pytest test_x86inductor_quantizer.py -k test_avg_pool2d_recipe
```



[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Add cat and avg_pool2d recipe into x86InductorQuantizer"


**Summary**
Add `cat` and `avg_pool2d` quantization recipe as input output share observer into `x86InductorQuantizer`.

**Test Plan**
```
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_same_inputs
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_single_input
clear && python -m pytest test_x86inductor_quantizer.py -k test_avg_pool2d_recipe
```



[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Add cat and avg_pool2d recipe into x86InductorQuantizer"


**Summary**
Add `cat` and `avg_pool2d` quantization recipe as input output share observer into `x86InductorQuantizer`.

**Test Plan**
```
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_same_inputs
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_single_input
clear && python -m pytest test_x86inductor_quantizer.py -k test_avg_pool2d_recipe
```



[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Add cat and avg_pool2d recipe into x86InductorQuantizer"


**Summary**
Add `cat` and `avg_pool2d` quantization recipe as input output share observer into `x86InductorQuantizer`.

**Test Plan**
```
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_same_inputs
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_single_input
clear && python -m pytest test_x86inductor_quantizer.py -k test_avg_pool2d_recipe
```



[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Add cat and avg_pool2d recipe into x86InductorQuantizer"


**Summary**
Add `cat` and `avg_pool2d` quantization recipe as input output share observer into `x86InductorQuantizer`.

**Test Plan**
```
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_same_inputs
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_single_input
clear && python -m pytest test_x86inductor_quantizer.py -k test_avg_pool2d_recipe
```



[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Add cat and avg_pool2d recipe into x86InductorQuantizer"


**Summary**
Add `cat` and `avg_pool2d` quantization recipe as input output share observer into `x86InductorQuantizer`.

**Test Plan**
```
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_same_inputs
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_single_input
clear && python -m pytest test_x86inductor_quantizer.py -k test_avg_pool2d_recipe
```



[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E] Add cat and avg_pool2d recipe into x86InductorQuantizer"


**Summary**
Add `cat` and `avg_pool2d` quantization recipe as input output share observer into `x86InductorQuantizer`.

**Test Plan**
```
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_same_inputs
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_single_input
clear && python -m pytest test_x86inductor_quantizer.py -k test_avg_pool2d_recipe
```



[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge
